### PR TITLE
(PUP-8127) change the error location output to be translationable

### PR DESCRIPTION
**Pull request by ericldelaney**
Created a common method for making a string of the the location of an error
Changed the error reporting and logging systems to use the common methods
Updated all the unit tests

### CODE REVIEW

**Comment by ericldelaney** [comment]
jenkins please test this

**Comment by ericldelaney** [comment]
jenkins please test this

**Comment by Magisus** [comment]
I do think some clarity is lost with this reformat, but I'm not sure how else to best make this language-agnostic. In terms of code verbosity, this iteration is definitely better than what was in there before, so that's a plus. And if we ever decide to end up changing the format, now it only needs to be changed in one spot, which I also like.

**Comment by jhelwig** [comment]
Yeah, I'm definitely with @Magisus here. I think a fair amount of clarity is lost by dropping `at line`, `in`, etc. but I'm not really seeing how to make this work in a way where translating it would make grammatical sense if they're kept.

**Comment by ericldelaney** [comment]
@Magisus @jhelwig I'm open to suggestions

**Comment by jhelwig** [comment]
Unfortunately, I don't have one. Right now it seems like it's a necessary evil.

The closest thing I can think of is how you deferred translation in a different PR you did before the holiday break (I'm forgetting which one off-hand), but I'm not sure that's actually reasonable to do here.

**Comment by ericldelaney** [comment]
jenkins please test this

**Comment by ericldelaney** [comment]
jenkins please test this

**Comment by ericldelaney** [comment]
jenkins please test this

File: lib/puppet/resource/catalog.rb [hunk 1]
```diff
@@ -591,12 +591,12 @@ def fail_on_duplicate_type_and_title(resource, title_key)
     return unless existing_resource = @resource_table[title_key]
 
     # If we've gotten this far, it's a real conflict
-    msg = "Duplicate declaration: #{resource.ref} is already declared"
-
-    msg << " in file #{existing_resource.file}:#{existing_resource.line}" if existing_resource.file and existing_resource.line
-
-    msg << "; cannot redeclare"
-
+    error_location_str = Puppet::Util::Errors.error_location(existing_resource.file, existing_resource.line)
```
Comment (ericldelaney) [inline lib/puppet/resource/catalog.rb]: This file is a new addition to the PR

File: lib/puppet/resource/catalog.rb [hunk 2]
```diff
@@ -591,12 +591,12 @@ def fail_on_duplicate_type_and_title(resource, title_key)
...
+    msg = if error_location_str.empty?
+            _("Duplicate declaration: %{resource} is already declared; cannot redeclare") % { resource: resource.ref }
+          else
+            _("Duplicate declaration: %{resource} is already declared; at %{error_location} cannot redeclare") % { resource: resource.ref, error_location: error_location_str }
```
Comment (jhelwig) [inline lib/puppet/resource/catalog.rb]: Looks like the `;` here was originally always before "cannot redeclare" (and after "at &lt;location&gt;" if it was there). Typo in moving it before the location?

Comment (ericldelaney) [reply lib/puppet/resource/catalog.rb]: Arg you are correct. Fixed

File: lib/puppet/util/errors.rb [hunk 3]
```diff
@@ -31,22 +31,51 @@ def adderrorcontext(error, other = nil)
...
   #
-  # @return [String] description of file and line
-  def error_context
-    if file and line
-      _(" at %{file}:%{line}") % { file: file, line: line }
+  # @return [String] description of file, line, and pos
+  def self.error_location(file, line=nil, pos=nil)
+    file = nil if (file.is_a?(String) && file.empty?)
+    if file and line and pos
+      _("(file: %{file}, line: %{line}, column: %{column})") % { file: file, line: line, column: pos }
```
Comment (Magisus) [inline lib/puppet/util/errors.rb]: 👍

**Pull request merged**